Allow combinations of `--change-time`, `--delete`, and `--edit` while correctly counting the number of entries affected

wren opened the pull request:
cc:
- #1640 
- #1639 

Related issues:
- Fixes #1696 
- Fixes #1666 
- Fixes #1639
- Fixes #1644 

### Todo
- [x] Add tests
- [x] Merge in `develop` branch


### Checklist

- [x] I have read the [contributing doc](https://github.com/jrnl-org/jrnl/blob/develop/CONTRIBUTING.md).
- [x] I have included a link to the relevant issue number.
- [x] I have checked to ensure there aren't other open [pull requests](../pulls)
  for the same issue.
- [x] I have written new tests for these changes, as needed.
<!--
NOTE: Your PR may not be reviewed if there are any failing tests. You can run
tests locally with `poe test` (see the contributing doc if you need help with
`poe`), or use our automated tests after you submit your PR.
-->

wren committed:
WIP

micahellison committed:
Remove ipdb and remove search mode conditional that added explicit tag search behavior

micahellison committed:
Fix failing change-time test by using same method signature as base journal class

micahellison committed:
Fix user input mock - was not appropriately checking return value

micahellison commented:
This might fix #1666. We'll need to add a test.

giuseppedandrea committed:
Update documentation on temporary files naming (#1673)

jrnl-bot committed:
Update changelog [ci skip]

Granddave committed:
Add documentation about information leaks in Vim/Neovim (#1674)

* Add documentation about using Vim/Neovim as editor
* Add documentation about information leaks in editors
* Spelling fix

---------

Co-authored-by: Jonathan Wren <jonathan@nowandwren.com>

Briscoooe committed:
Don't save templated journal entries if the received raw text is the same as the template itself (#1653)

jrnl-bot committed:
Update changelog [ci skip]

micahellison commented:
We'll need to add some tests for:
- [x] combinations of action arguments (edit, change-time, delete)
- [x] combinations of stderr output messaging (entries found, entries modified)
- [x] might need some tests for 0 vs 1 vs multiple entries found and/or modified
- [ ] unit tests for the new private controller functions

We'll need to come up with a better name for this PR too. It changes some messaging behavior and we'll want that reflected in the changelog.

wren committed:
Clean up controller

- streamline `run` function in `controller.py`
- add debug logging
- fix unnecessary import of Journal class (only needed for typing)
- standardize summary display across different actions

micahellison committed:
Add currently-failing test conditions for count messages when changing time and deleting

micahellison commented:
I noticed some problems with incorrect count messages while testing this out today, so I added some failing tests for those issues.

micahellison committed:
Don't show summary if no entries found and prevent extra line break when no entries found by short-circuiting display method

micahellison committed:
Track found entry count and remove incorrect modified stat logic

micahellison committed:
Track journal entry deletion consistently

micahellison committed:
Remove unneeded exception when editor is empty and fix test that was testing incorrect message

micahellison committed:
Correct entry edit modified count test

micahellison committed:
Track modification of entries with --change-time

micahellison committed:
Preserve existing behavior when editor is empty but make the message more clear

micahellison committed:
Reconcile tests with new error message when clearing editor in edit mode

micahellison committed:
Fix unit test that did not account for new short-circuit when displaying journal with no entries

micahellison committed:
Fix other unit test that did not account for short-circuit

micahellison commented:
This is green but not finished. We still need to add the tests I mentioned in my comment last week.

We also changed the messaging for when you delete all entries in the editor. Instead of the `No entry to save, because no text was received` message, we're showing this warning:
```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃  No text received from editor. Were you trying to delete all the entries?  ┃
┃                                                                            ┃
┃  This seems a bit drastic, so the operation was cancelled.                 ┃
┃                                                                            ┃
┃  To delete all entries, use the --delete option.                           ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
```

micahellison committed:
Add found/modified counts to edit tests

micahellison committed:
Add tests for found count with -n equivalent argument

micahellison committed:
Test combinations of found/deleted messages when using --delete

micahellison committed:
Add tests for counting combinations of action arguments (change-time, edit, delete) and for change-time counts. Some are failing and should be investigated

micahellison commented:
I added some failing tests that we should look at. I'm not 100% certain what the added/found/deleted counts should be with certain combinations of actions, but I still think they are wrong now in certain situations.

micahellison committed:
Remove extraneous comment in test

micahellison committed:
Track added/deleted counts in a register in the Journal class instead of attempting to infer it via controller counting

micahellison committed:
run poe format

micahellison commented:
I finally decided to just keep a register of deleted and added entries in the Journal class. We're already tracking modified entries in its member class Entry so this is not a significant deviation, and now the counts reflect exactly what jrnl did to the journal, even when combining all three actions (change-time, delete, and edit).

This change also has a nice side effect of moving all counting logic outside of the controller class, which has no business doing that anyway. In the long run, I'd like to separate this logic from the data representation of the journal, but I think that's outside of the scope of this PR.

micahellison commented:
I still want to add more tests to this. The combination tests are a good start, but I want to test them with search criteria as well.

wren committed:
Merge branch 'develop' into mode-actions-1639

Conflicts:
  CHANGELOG.md
  docs/privacy-and-security.md
  jrnl/controller.py
  tests/bdd/features/change_time.feature

wren committed:
run format

wren committed:
add encrypted to tests

wren committed:
rename test file

wren committed:
add encrypted to tests for change-time

wren committed:
Merge branch 'develop' into mode-actions-1639

Conflicts:
  jrnl/controller.py

wren committed:
fix merge conflict typo

wren committed:
'write mode' -> 'append mode' in more places

wren committed:
run format
